src/endpoints.rs

```diff
@@ -207,3 +208,63 @@ pub async fn get_region(
             .map(|trainee| trainee.region.clone()),
     }))
 }
+
+#[derive(Serialize)]
+pub struct AttendanceResponse {
+    courses:
+        IndexMap<String, IndexMap<String, IndexMap<String, IndexMap<String, Vec<Attendance>>>>>,
+}
+
+pub async fn fecth_attendance(
+    session: Session,
+    State(server_state): State<ServerState>,
+    OriginalUri(original_uri): OriginalUri,
+) -> Result<Json<AttendanceResponse>, Error> {
+    let all_courses = &server_state.config.courses;
+    let sheets_client = sheets_client(&session, server_state.clone(), original_uri.clone()).await?;
+    let github_org = &server_state.config.github_org;
+    let octocrab = octocrab(&session, &server_state, original_uri).await?;
+    let mut courses: IndexMap<
+        String,
+        IndexMap<String, IndexMap<String, IndexMap<String, Vec<Attendance>>>>,
+    > = IndexMap::new();
+    for (course_name, course_info) in all_courses {
+        let mut course_batches: IndexMap<
+            String,
+            IndexMap<String, IndexMap<String, Vec<Attendance>>>,
+        > = IndexMap::new();
+        for batch_name in course_info.batches.keys() {
+            let course_schedule = server_state
+                .config
+                .get_course_schedule_with_register_sheet_id(course_name.clone(), &batch_name)
+                .ok_or_else(|| Error::Fatal(anyhow::anyhow!("Course not found: {course_name}")))?;
+            let mut modules: IndexMap<String, IndexMap<String, Vec<Attendance>>> = IndexMap::new();
+            let course = course_schedule
+                .with_assignments(&octocrab, github_org)
+                .await?;
+            let register_info = get_register(
+                sheets_client.clone(),
+                course.register_sheet_id.clone(),
+                course.start_date,
+                course.end_date,
+            )
+            .await?;
```

Comment:
Because the `await` here is inside a loop, this will look up one register at a time, rather than looking them up in parallel, which slightly defeats the point of this being async.

We probably don't care about performance of this that much, so it's not a big deal, but sometimes we use a pattern of listing all the courses, then lookup up all of their registers in parallel (using something like [`join_all`](https://docs.rs/futures/latest/futures/future/fn.join_all.html)) and then merging the two lists.